### DISCUSSION TITLE
Any way to set keep alive for the underlying asyncssh transport?

### DISCUSSION THREAD

**Discussion by lamoni** [scrapli/scrapli_netconf#96] [Q&A]
@carlmontanari love this module you've put together, much appreciated! I was curious if there is a way to set an ssh keepalive interval when asyncssh is used as the transport

**Comment by carlmontanari** [comment]
Hey @lamoni thanks a bunch!

Not at the moment (unless asyncssh is already magically picking it up from ssh config file -- which is totally possible because Ron/asyncssh is 🔥 ). *But* we do have a means to expose those type of flags -- see [here](https://github.com/carlmontanari/scrapli/blob/master/scrapli/transport/plugins/asyncssh/transport.py#L50-L69) and [here](https://github.com/carlmontanari/scrapli/blob/master/scrapli/transport/plugins/asyncssh/transport.py#L168-L169) -- so I think we can add things fairly easily. If you can show some examples in "native" asyncssh on what you are thinking I suspect we could get it added. I'd be curious what devices you are connecting to that are actually honoring keep alive interval? In my experience w/ cisco/arista things it seems that mostly things like that/control persist/etc are ignored (though its been a while since I investigated!)

Carl

**Reply by lamoni** [reply]
Such quick response, appreciate it. 

Sorry I should have clarified, I mean from the client side (or please let me know if I'm misunderstanding your question). I'm using scrapli_netconf with asyncssh transport to keep persistent netconf connections to some Junipers. Certain versions seem to have issues with killing the netconf connection even if the connection has been idle for less than the configured idle-timeout on the device (sigh), so I'm hoping to work around it by sending shorter interval keepalives. No idea if that will actually workaround the issue but was hoping to test it quickly :)

e.g. from asyncssh's docs:

keepalive_interval (see [Specifying time intervals](https://asyncssh.readthedocs.io/en/latest/api.html#specifyingtimeintervals)) – (optional) The time in seconds to wait before sending a keepalive message if no data has been received from the server. This defaults to 0, which disables sending these messages.

So I'm hoping to have the client (us) send keepalive messages, equivalent to ssh option "ServerAliveInterval" I think?

Equivalent code for asyncssh might look like:

```
options = asyncssh.SSHClientConnectionOptions(keepalive_interval=10) # consider testing tcp_keepalive option as well?
async with asyncssh.connect('lab1-test-device', username='blah', password='blah', options=options) as conn:
```

**Reply by carlmontanari** [reply]
Ok, gotcha. You may be able to test this out easily like: 

```python
        device = {
            "host": "blah",
            "transport_options": {
                "asyncssh": {
                    "keepalive_interval": 10,
                }
            },
            "platform": "blah"
        }
```

From a super quick pass at the docs and looking back at the asyncssh transport plugin, it *looks* like that will pass that kwarg straight to asyncssh (which it looks like accepts these as kwargs just like encryption args and the like)... if you can give that a crack and let us know if it works that'd be great. I can try to tickle it this weekend a bit too.
